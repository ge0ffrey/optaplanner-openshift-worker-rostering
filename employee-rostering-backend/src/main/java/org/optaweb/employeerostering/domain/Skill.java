/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.optaweb.employeerostering.domain;

import java.util.Objects;

public class Skill extends AbstractPersistable {

    private final Integer tenantId;
    private final String name;

    public Skill(Integer tenantId, String name) {
        this.tenantId = Objects.requireNonNull(tenantId);
        this.name = Objects.requireNonNull(name);
    }

    public Integer tenantId() {
        return tenantId;
    }

    public String name() {
        return name;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Skill skill = (Skill) o;
        return tenantId.compareTo(skill.tenantId) == 0 &&
                name.compareTo(skill.name) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(tenantId, name);
    }

    @Override
    public String toString() {
        return "Skill{" +
                "tenantId=" + tenantId +
                "name=" + name +
                '}';
    }
}
