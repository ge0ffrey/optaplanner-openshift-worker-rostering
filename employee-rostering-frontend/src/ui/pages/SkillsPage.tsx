/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import DataTable from '../components/DataTable';
import SkillRestServiceClient from '../../services/SkillRestServiceClient';

const skillRestService = new SkillRestServiceClient();

function getSkills(callback: (_: string[][]) => void): void {
  skillRestService.getSkillList(1).then((res) => {
    callback(res.map(skill => [skill.name]));
  });
}

const SkillsPage = () => {
  return <DataTable title={'Skills'} columnTitles={['Name']} dataSupplier={getSkills}></DataTable>;
};

export default SkillsPage;
