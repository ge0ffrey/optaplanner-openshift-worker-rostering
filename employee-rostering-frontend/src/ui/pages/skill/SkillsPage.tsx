/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import {DataTable, DataTableProps, ReadonlyPartial, PropertySetter, Sorter} from 'ui/components/DataTable';
import {skillOperations} from 'store/skill';
import Skill from 'domain/Skill';
import { AppState } from 'store/types';
import { TextInput, Text } from '@patternfly/react-core';
import { connect } from 'react-redux';
import { Predicate } from 'ui/components/FilterComponent';
import { stringSorter } from 'util/CommonSorters';
import { stringFilter } from 'util/CommonFilters';

interface StateProps extends DataTableProps<Skill> {
  tenantId: number;
}

const mapStateToProps = ({ tenantData, skillList }: AppState): StateProps => ({
  title: "Skills",
  columnTitles: ["Name"],
  tableData: skillList.skillList,
  tenantId: tenantData.currentTenantId
}); 

export interface DispatchProps {
  addSkill: typeof skillOperations.addSkill;
  updateSkill: typeof skillOperations.updateSkill;
  removeSkill: typeof skillOperations.removeSkill;
}

const mapDispatchToProps: DispatchProps = {
  addSkill: skillOperations.addSkill,
  updateSkill: skillOperations.updateSkill,
  removeSkill: skillOperations.removeSkill
};

export type Props = StateProps & DispatchProps;

export class SkillsPage extends DataTable<Skill, Props> {
  constructor(props: Props) {
    super(props);
    this.addData = this.addData.bind(this);
    this.updateData = this.updateData.bind(this);
    this.removeData = this.removeData.bind(this);
  }

  displayDataRow(data: Skill): JSX.Element[] {
    return [<Text key={0}>{data.name}</Text>];
  }

  getInitialStateForNewRow(): Partial<Skill> {
    return {};
  }

  editDataRow(data: ReadonlyPartial<Skill>, setProperty: PropertySetter<Skill>): JSX.Element[] {
    return [<TextInput
      key={0}
      name="name"
      aria-label="Name"
      defaultValue={data.name}
      onChange={(value) => setProperty("name", value)}
    />];
  }
  
  isValid(data: ReadonlyPartial<Skill>): data is Skill {
    return data.name !== undefined && 
      data.name.length > 0;
  }

  getFilter(): (filter: string) => Predicate<Skill> {
    return stringFilter((skill) => skill.name);
  }

  getSorters(): (Sorter<Skill> | null)[] {
    return [stringSorter(s => s.name)];
  }

  updateData(data: Skill): void {
    this.props.updateSkill(data);
  }
  
  addData(data: Skill): void {
    this.props.addSkill({...data, tenantId: this.props.tenantId});
  }

  removeData(data: Skill): void {
    this.props.removeSkill(data);
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(SkillsPage);
