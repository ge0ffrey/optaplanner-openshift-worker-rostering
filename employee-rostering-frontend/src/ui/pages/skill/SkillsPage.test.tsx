/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { SkillsPage, Props } from './SkillsPage';
import Skill from 'domain/Skill';

describe('Skills page', () => {
  it('should render correctly with no skills', () => {
    const skillsPage = shallow(<SkillsPage {...noSkills} />);
    expect(toJson(skillsPage)).toMatchSnapshot();
  });

  it('should render correctly with a few skills', () => {
    const skillsPage = shallow(<SkillsPage {...twoSkills} />);
    expect(toJson(skillsPage)).toMatchSnapshot();
  });

  it('should render the viewer correctly', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const skill = {name: "Skill", tenantId: 0, id: 1, version: 0};
    const viewer = shallow(skillsPage.renderViewer(skill));
    expect(toJson(viewer)).toMatchSnapshot();
  });

  it('should render the editor correctly', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const skill = {name: "Skill", tenantId: 0, id: 1, version: 0};
    const editor = shallow(skillsPage.renderEditor(skill));
    expect(toJson(editor)).toMatchSnapshot();
  });

  it('should return a skill with the tenantId but no name, id or version on createNewDataInstance', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const skill = skillsPage.createNewDataInstance();
    expect(skill.tenantId).toEqual(twoSkills.tenantId);
    expect(skill.name).toEqual("");
    expect(skill.id).toBeUndefined();
    expect(skill.version).toBeUndefined();
  });

  it('should call addSkill on addData', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const skill = {name: "Skill", tenantId: 0};
    skillsPage.addData(skill);
    expect(twoSkills.addSkill).toBeCalled();
    expect(twoSkills.addSkill).toBeCalledWith(skill);
  });

  it('should call updateSkill on updateData', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const skill = {name: "Skill", tenantId: 0, id: 1, version: 0};
    skillsPage.updateData(skill);
    expect(twoSkills.updateSkill).toBeCalled();
    expect(twoSkills.updateSkill).toBeCalledWith(skill);
  });

  it('should call removeSkill on removeData', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const skill = {name: "Skill", tenantId: 0, id: 1, version: 0};
    skillsPage.removeData(skill);
    expect(twoSkills.removeSkill).toBeCalled();
    expect(twoSkills.removeSkill).toBeCalledWith(skill);
  });

  it('should treat empty name as invalid', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const components: Skill = {tenantId: 0, name: ""};
    const result = skillsPage.isValid(components);
    expect(result).toEqual(false);
  });

  it('should treat non-empty name as valid', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const components: Skill = {tenantId: 0, name: "Skill"};
    const result = skillsPage.isValid(components);
    expect(result).toEqual(true);
  });
});

const noSkills: Props = {
  tenantId: 0,
  title: "Skills",
  columnTitles: ["Name"],
  tableData: [],
  addSkill: jest.fn(),
  updateSkill: jest.fn(),
  removeSkill: jest.fn()
};

const twoSkills: Props = {
  tenantId: 0,
  title: "Skills",
  columnTitles: ["Name"],
  tableData: [{
    id: 0,
    version: 0,
    tenantId: 0,
    name: "Skill 1"
  },
  {
    id: 1,
    version: 0,
    tenantId: 0,
    name: "Skill 2"
  }],
  addSkill: jest.fn(),
  updateSkill: jest.fn(),
  removeSkill: jest.fn()
};