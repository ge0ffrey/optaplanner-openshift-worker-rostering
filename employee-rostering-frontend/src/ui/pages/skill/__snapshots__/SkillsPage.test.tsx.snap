// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skills page should render correctly with a few skills 1`] = `
<DataTable
  columnTitles={
    Array [
      "Name",
    ]
  }
  createRow={[Function]}
  rowDataToRow={[Function]}
  tableData={
    Array [
      Object {
        "id": 0,
        "name": "Skill 1",
        "tenantId": 0,
        "version": 0,
      },
      Object {
        "id": 1,
        "name": "Skill 2",
        "tenantId": 0,
        "version": 0,
      },
    ]
  }
  title="Skills"
/>
`;

exports[`Skills page should render correctly with no skills 1`] = `
<DataTable
  columnTitles={
    Array [
      "Name",
    ]
  }
  createRow={[Function]}
  rowDataToRow={[Function]}
  tableData={Array []}
  title="Skills"
/>
`;
